sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNullAwareHashJoinExec.scala

```diff
@@ -0,0 +1,163 @@
+
+  // TODO support multi column not in subquery in future.
+  // See. http://www.vldb.org/pvldb/vol2/vldb09-423.pdf Section 6
+  // multi-column null aware anti join is much more complicated than single column ones.
+  require(right.output.length == 1, "not in subquery hash join optimize only single column.")
+  require(joinType == LeftAnti, "joinType must be LeftAnti.")
+  require(buildSide == BuildRight, "buildSide must be BuildRight.")
+  require(SQLConf.get.notInSubqueryHashJoinEnabled,
+    "notInSubqueryHashJoinEnabled must turn on for BroadcastNullAwareHashJoinExec.")
+
+  override def leftKeys: Seq[Expression] = Nil
+  override def rightKeys: Seq[Expression] = Nil
+
+  override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+
+  private val (streamed, broadcast) = (left, right)
+
+  override def simpleStringWithNodeId(): String = {
+    val opId = ExplainUtils.getOpId(this)
+    s"$nodeName $joinType ${buildSide} ($opId)".trim
+  }
+
+  override def requiredChildDistribution: Seq[Distribution] = {
+    UnspecifiedDistribution :: BroadcastDistribution(IdentityBroadcastMode) :: Nil
+  }
+
+  private[this] def genResultProjection: UnsafeProjection = {
+    UnsafeProjection.create(output, output)
+  }
+
+  override def output: Seq[Attribute] = {
+    left.output
+  }
+
+  private def buildParams(buildSideRows: Array[InternalRow]): NotInSubqueryHashJoinParams = {
+    val (_, leftAttr, rightAttr) =
+      NotInSubqueryConditionPattern.singleColumnPatternMatch(condition)
+
+    val leftNotInKeyFound = left.output.exists(_.semanticEquals(leftAttr))
+    val (streamedKey, streamedKeyIndex) = if (leftNotInKeyFound) {
+      (leftAttr, AttributeSeq(left.output).indexOf(leftAttr.exprId))
+    } else {
+      (rightAttr, AttributeSeq(left.output).indexOf(rightAttr.exprId))
+    }
+
+    NotInSubqueryHashJoinParams(
+      HashedRelation(buildSideRows.iterator,
+        BindReferences.bindReferences[Expression](
+          Seq(right.output.head), AttributeSeq(right.output)),
+        buildSideRows.length),
+      buildSideRows.exists(row => row.isNullAt(0)),
+      buildSideRows.isEmpty,
+      streamedKey,
+      streamedKeyIndex
+    )
+  }
+
+  private def leftAntiNullAwareJoin(
+      relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+    val params = buildParams(relation.value)
+    streamed.execute().mapPartitionsInternal { streamedIter =>
+      if (params.buildSideIsEmpty) {
+        streamedIter
+      } else if (params.buildSideIsNullExists) {
+        Iterator.empty
+      } else {
+        val keyGenerator = UnsafeProjection.create(
+          BindReferences.bindReferences[Expression](
+            Seq(params.streamedSideKey),
+            AttributeSeq(left.output))
+        )
+        streamedIter.filter(row => {
+          val streamedRowIsNull = row.isNullAt(params.streamedSideKeyIndex)
+          val lookupRow: UnsafeRow = keyGenerator(row)
```

Comment:
nit: Should this be renamed to lookupKey ?

Reply:
done